system/codesigner/gpgcodesigner.cpp

```
@@ -0,0 +1,250 @@
+
+    StringBuffer buf;
+    Owned<ISimpleReadStream> pipeReader = pipe->getErrorStream();
+    readSimpleStream(buf, *pipeReader);
+    DBGLOG("GPG %d %s", retcode, buf.str());
+    if (retcode)
+    {
+        errmsg.setf("gpg verify failed: %s (%d)", buf.str(), retcode);
+    }
+    else
+    {
+        const char * sigprefix = "Good signature from \"";
+        const char * const s = buf.str();
+        const char * match = strstr(s, sigprefix);
+        if (match)
+        {
+            match += strlen(sigprefix);
+            const char * const end = strchr(match, '\"');
+            if (end)
+            {
+                signer.append(end-match, match);
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+bool GpgCodeSigner::hasSignature(const char *text) const
+{
+    return startsWith(text, signatureMsgHeader);
+}
+
+StringBuffer &GpgCodeSigner::stripSignature(const char *text, StringBuffer &unsignedText) const
+{
+    if (!hasSignature(text))  // no signature -> return unchanged
+        return unsignedText.set(text);
+
+    const char *head = text;
+    head += strlen(signatureMsgHeader);     // skip header
+    while ((head = strchr(head, '\n')) != nullptr)
+    {
+        head++;
+        if (*head=='\n')
+        {
+            head++;
+            break;
+        }
+        else if (*head=='\r' && head[1]=='\n')
+        {
+            head += 2;
+            break;
+        }
+    }
+
+    if (!head)
+        return unsignedText.set(text);
+
+    const char *tail = strstr(head, signatureBegin);
+    if (!tail)
+        return unsignedText.set(text);
+    return unsignedText.append(tail-head, head);
+}
+
+const char* skipn(const char* str, char c, int n)
+{
+    for (int i = 0; i < n && str && *str; i++)
+    {
+        str = strchr(str, c);
+        if (!str)
+            break;
+        str++;
+    }
+    return str;
+}
+
+StringArray &GpgCodeSigner::getUserIds(StringArray &userIds, StringBuffer &errmsg)
```

Not sure why errmsg is passed as a parameter - it will never be used by the caller.

---

I've changed verifySignature to throw exceptions instead of returning error messages in errmsg.